### ISSUE TITLE
Stalled ranks and deadlock when Horovod distribute train finished and start evaluation 

### ISSUE THREAD

**Issue by tiandongtao** [horovod/horovod#3781]
**Environment:**
1. Framework: (TensorFlow2.1.0, Keras, PyTorch1.4.0, MXNet1.6.0)
2. Framework version:
3. Horovod version:0.19.2
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:3.6
8. Spark / PySpark version:
9. Ray version:
10. OS and version:
11. GCC version:
12. CMake version:

**Checklist:**
1、Did you search issues to find if somebody asked this question before? Yes
2、If your question is about hang, did you read this doc? Yes
3、If your question is about docker, did you read this doc? Yes
4、Did you check if you question is answered in the troubleshooting guide? Yes
**Bug report:**
I use the command to run job
mpirun -np 32 --allow-run-as-root -H ${ip[0]}:4,${ip[1]}:4,${ip[2]}:4,${ip[3]}:4,${ip[4]}:4,${ip[5]}:4,${ip[6]}:4,${ip[7]}:4 python ./run.py

train and evaluation：

if is_train:
	random.shuffle(train_files)
	estimator.train(input_fn=make_input_fn(train_files, batch_size=Configs['batch_size'], num_epochs=Configs['epochs'],shuffle=Configs['shuffle'], num_workers=hvd.size(), hvd_index=hvd.rank()), steps=None, hooks=[bcast_hook])

if is_eval:
   random.shuffle(eval_files)
   results = estimator.evaluate(input_fn=make_input_fn(eval_files, batch_size=Configs['batch_size'], num_epochs=1, shuffle=Configs['shuffle'], num_workers=hvd.size(), hvd_index=hvd.rank()), steps=None,hooks=[bcast_hook])

when train finished and start evaluation ,the job deadlock,shows this:

W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
Stalled ranks:
[DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_duration_bucketized_embedding_item_duration_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_0, DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_duration_bucketized_embedding_item_duration_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_1_0, DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_rtype_stay_duration_15d_bucketized_embedding_item_rtype_stay_duration_15d_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_0, DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_rtype_stay_duration_15d_bucketized_embedding_item_rtype_stay_duration_15d_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_1_0, DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_rtype_stay_duration_7d_bucketized_embedding_item_rtype_stay_duration_7d_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_0, DistributedAdamOptimizer_Allreduce/HorovodAllgather_gradients_feature_shallow_tower_1_input_layer_item_rtype_stay_duration_7d_bucketized_embedding_item_rtype_stay_duration_7d_bucketized_embedding_weights_embedding_lookup_sparse_embedding_lookup_grad_Reshape_1_0 ...]

**Comment by tiandongtao** [comment]
who can reply this problem? @maxhgerlach
